Add neural network modules, tests, and test utils

kevinchern opened the pull request.

thisac commented on dwave/plugins/torch/nn.py:
```diff
@@ -0,0 +1,118 @@
+            x (torch.Tensor): the input tensor.
+
+        Returns:
+            torch.Tensor: the linearly-transformed tensor of `x`.
```
```suggestion
        """Apply a linear transformation to the input variable ``x``.

        Args:
            x (torch.Tensor): The input tensor.

        Returns:
            torch.Tensor: The linearly-transformed tensor of ``x``.
```

thisac commented on dwave/plugins/torch/nn.py:
```diff
@@ -0,0 +1,118 @@
+
+class LinearBlock(nn.Module):
+    @store_config
+    def __init__(self, din, dout, p) -> None:
```
Missing typehints.

thisac commented on dwave/plugins/torch/nn.py:
```diff
@@ -0,0 +1,118 @@
+import torch
+from torch import nn
+
+
```
Don't forget to add an `__all__` here. See e.g., [here](https://github.com/dwavesystems/dwave-pytorch-plugin/blob/aa981186b1582c133019bdb7d0311b31f5228113/dwave/plugins/torch/models/boltzmann_machine.py#L46).

thisac commented on dwave/plugins/torch/nn.py:
```diff
@@ -0,0 +1,118 @@
+            nn.ReLU(),
+            nn.LayerNorm(dout),
+            linear_2,
+        )
```
Is there a reason for not just doing the following? Since each linear layer is only used once it makes more sense to create them in place, as the other layers.
```suggestion
        self.block = nn.Sequential(
            nn.LayerNorm(din),
            nn.Linear(din, dout),
            nn.Dropout(p),
            nn.ReLU(),
            nn.LayerNorm(dout),
            nn.Linear(dout, dout),
        )
```
or even adding in the skip, and removing it from forward
```suggestion
        self.block = nn.Sequential(
            nn.LayerNorm(din),
            nn.Linear(din, dout),
            nn.Dropout(p),
            nn.ReLU(),
            nn.LayerNorm(dout),
            nn.Linear(dout, dout),
            SkipLinear(din, dout),
        )
```

thisac commented on dwave/plugins/torch/nn.py:
```diff
@@ -0,0 +1,118 @@
+    def __init__(self, din, dout) -> None:
+        """Applies a linear transformation to the incoming data: :math:`y = xA^T`.
+
+        This module is identity when `din == dout`, otherwise, it is a linear transformation, i.e.,
```
```suggestion
        This module is identity when ``din == dout``, otherwise, it is a linear transformation, i.e.,
```

VolodyaCO replied:
Is this what makes it a "Skip" linear instead of a plain linear module? I find this very counter-intuitive: that having the same output dimension performs an identity transformation instead of a linear transformation (like an endomorphism).

kevinchern commented on dwave/plugins/torch/nn.py:
```diff
@@ -0,0 +1,118 @@
+    return wrapper
+
+
+class Identity(nn.Module):
```
uh LOL. How did I miss that 😆

VolodyaCO commented on dwave/plugins/torch/nn.py:
```diff
@@ -0,0 +1,118 @@
+            p (float): Dropout probability.
+        """
+        super().__init__()
+        self.skip = SkipLinear(din, dout)
```
The input of forward is `x`, if you add skip to the end of sequential, it would be performing the operation `skip(block(x))`, but what is wanted is `skip(x) + block(x)`.

VolodyaCO reviewed (commented):
Left a question about `SkipLinear`

anahitamansouri commented on dwave/plugins/torch/nn.py:
```diff
@@ -0,0 +1,127 @@
+        config['module_name'] = self.__class__.__name__
+        self.config = MappingProxyType(config)
+
+        fn(self, *args, **kwargs)
```
I was wondering if you should return `fn(...)` or not?

anahitamansouri commented on releasenotes/notes/add-nn-modules-c29a092140eacbe1.yaml:
```diff
@@ -0,0 +1,6 @@
+features:
+  - Add the python module ``dwave.plugins.torch.nn`` containing commonly-used neural network modules
+    and patterns used to build more complex architectures.
+  - Add ``LinearBlock`` and ``SkipLinear` modules.
```
```suggestion
  - Add ``LinearBlock`` and ``SkipLinear`` modules.
```

anahitamansouri commented on dwave/plugins/torch/nn.py:
```diff
@@ -0,0 +1,127 @@
+        return self.linear(x)
+
+
+class LinearBlock(nn.Module):
```
Could you please motivate the architecture? I assume you have been using this for something and you needed this in the plugin. Right?

anahitamansouri reviewed (commented):
Thanks very much Kevin for the great work. I have left my comments.

kevinchern commented:
RE @thisac : restructured the file into (sub)modules by pattern-matching pytorch's structure.
RE @VolodyaCO : configs are now stored recursely for nested models :D. reference for resnet added in the docstring
RE @anahitamansouri : fixed typos and added reference for resnet in docstring

thisac commented on dwave/plugins/torch/nn/modules/linear.py:
```diff
@@ -0,0 +1,98 @@
+    """A linear block consisting of normalizations, linear transformations, dropout, relu, and a skip connection.
+
+    The module is composed of (in order):
+    1. a first layer norm,
```
I _think_ Sphinx requires the first element in a list to have an empty line above to render properly (in case we render docs).

```suggestion
    The module is composed of (in order):

    1. a first layer norm,
```

thisac commented on dwave/plugins/torch/nn/modules/linear.py:
```diff
@@ -0,0 +1,98 @@
+        dout (int): Size of each output sample.
+        p (float): Dropout probability.
+    """
+    @store_config
```
```suggestion
    """

    @store_config
```

thisac commented on dwave/plugins/torch/nn/modules/linear.py:
```diff
@@ -0,0 +1,118 @@
+        )
+
+    def forward(self, x) -> torch.Tensor:
+        """Transforms the input `x` with the modules.
```
Missed this one!

thisac commented on tests/test_nn.py:
```diff
@@ -0,0 +1,70 @@
+
+class TestUtils(unittest.TestCase):
+
+    def test_store_config(self):
```
Should split the simple case and nested case into two separate tests.

thisac commented on tests/test_nn.py:
```diff
@@ -0,0 +1,70 @@
+        module_1 = InnerModel(a=123, x=5)
+        module_2 = InnerModel(a="second", y="lol")
+        model = OuterModel(module_1, module_2)
+        self.assertDictEqual(dict(model.config),
```
Does this work well comparing nested `MappingProxyType` objects? I wonder if you'd need to test e.g.,
```python
self.assertDictEqual(dict(model.config[module_1]), {"a": 123, "b": 1:, "x": 5, "y": "hello"}
self.assertDictEqual(dict(model.config[module_2]), {"a": "second", "b": 1:, "x": 4, "y": "lol"}
```
separately as well.

thisac commented on tests/test_nn.py:
```diff
@@ -0,0 +1,70 @@
+        dout = 99
+        model = SkipLinear(din, dout)
+        self.assertTrue(model_probably_good(model, (din,), (dout, )))
+        with self.subTest("Check identity for `din == dout`"):
```
This seems like it should be a stand-alone test.

kevinchern replied:
Why?

thisac replied:
Because testing the case when `din == dout` is a separate case to the one where `din != dout`. Having them as separate tests makes it a bit clearer and easier to parse the results. Also, there's very little code deduplication by having it in one test. 

Each test case should optimally be run separately, either as subtests, parameterized tests, or completely stand-alone.

thisac replied:
Yeah, that makes sense to me.

thisac commented on tests/test_test_utils.py:
```diff
@@ -0,0 +1,143 @@
+from tests import utils
+
+
+class TestUtils(unittest.TestCase):
```
Not sure I like the `test_test_utils.py` name. Maybe better to call this something different. Perhaps "helper_functions"? Not sure. Or simply call the file `test_utilities.py` and have it contain both the utility function _and_ it's tests. Or rename the other `test_utils.py` to `test_torch_utils.py`.

thisac commented on tests/test_test_utils.py:
```diff
@@ -0,0 +1,143 @@
+                return torch.ones(5)
+        model = MyModel(**kwargs)
+        self.assertTrue(utils.has_correct_config(model))
+        self.assertFalse(utils.has_correct_config(torch.nn.Linear(5, 3)))
```
This gets run twice (identically) due to the parameterization. I'd put it as a separate test.

thisac commented on tests/utils.py:
```diff
@@ -0,0 +1,123 @@
+        bool: Indicator for whether the model meets the three conditions above.
+    """
+    bs = 100
+    x = torch.randn((bs, ) + shape_in)
```
Never too fond of randomness in tests, if it can be avoided. 🫤

kevinchern replied:
I think hard-coded are more likely to miss bugs than this though. Do u have a suggestion for what hard-coded value would be useful here? (Or alternatives to this test)

kevinchern replied:
Or maybe this is not the right way to test these modules...

thisac commented on tests/utils.py:
```diff
@@ -0,0 +1,123 @@
+    return (x == 0).float().any()
+
+
+def bounded_in_plus_minus_one(x: torch.Tensor):
```
```suggestion
def bounded_in_plus_minus_one(x: torch.Tensor) -> bool:
```

kevinchern commented on dwave/plugins/torch/nn/modules/linear.py:
```diff
@@ -0,0 +1,98 @@
+        super().__init__()
+        self._skip = SkipLinear(din, dout)
+        dhid = max(din, dout)
+        self.block = nn.Sequential(
```
oh yeah thanks! that's definitely a mistake lol

kevinchern commented on tests/utils.py:
```diff
@@ -0,0 +1,123 @@
+import torch
+
+
+def model_probably_good(
```
Not sure. I only ever use it as unit "tests", more like sanity checks, for modules

kevinchern commented on tests/test_test_utils.py:
```diff
@@ -0,0 +1,143 @@
+        # Activate
+        self.assertFalse(utils.probably_unconstrained(x.sigmoid()))
+        self.assertFalse(utils.probably_unconstrained(x.relu()))
+        self.assertFalse(utils.probably_unconstrained(x.tanh()))
```
Huh, why? Isn't this standard practice?

kevinchern replied:
sry missed this one. for posterity: wouldn't this reasoning imply every test can only have a single assertion?
the test is for one behaviour on three input values of same shape and dtype, which seems reasonable to be in a single test

kevinchern commented on tests/test_nn.py:
```diff
@@ -0,0 +1,70 @@
+
+        model = MyModel(a=123, x=5)
+        self.assertDictEqual(dict(model.config),
+                             {"a": 123, "b": 1, "x": 5, "y": "hello", "module_name": "MyModel"})
```
what case do u have in mind?

thisac replied:
Not necessarily necessary here, but currently you're only testing for `MyModel(a=123, x=5)`. Other cases could be with different types (are only strings/ints allowed?), args instead of kwargs (or combinations thereof). Also, separate tests could be checking that storing a config with empty signature (only a name) works, only with default values, etc.

kevinchern commented:
@thisac  updated again based on your feedback on best practices. Thanks!!

1) added more tests for the store_config case and separated them as recommended
2) split linear tests as recommended

thisac commented on dwave/plugins/torch/nn/modules/linear.py:
```suggestion
        """Transforms the input ``x`` with the modules.
```

thisac commented on tests/test_helper_functions.py:
```diff
@@ -0,0 +1,148 @@
+
+                def forward(self, x):
+                    return 2*x
+            self.assertTrue(helper_functions.model_probably_good(MyModel("hello"), (500, ), (500,)))
```
To a point made earlier, this test could be simplified a bit:
```suggestion
        class MyModel(torch.nn.Module):
            @store_config
            def __init__(self, a, b=2, *, x=4, y=5):
                super().__init__()

        with self.subTest("Model should be good"):
            MyModel.forward = lambda self, x: 2 * x
            self.assertTrue(helper_functions.model_probably_good(MyModel("hello"), (500, ), (500,)))
```
but I'll leave this just as an option. Up to you whether you prefer this or the former.

thisac commented on tests/test_helper_functions.py:
```diff
@@ -0,0 +1,148 @@
+from tests import helper_functions
+
+
+class TestUtils(unittest.TestCase):
```
```suggestion
class TestHelperFunctions(unittest.TestCase):
```
